crates/noirc_evaluator/src/ssa_refactor/ir/cfg.rs

```diff
@@ -86,27 +86,149 @@ impl ControlFlowGraph {
         successor_node.predecessors.insert(from);
     }
 
-    /// Get an iterator over the CFG predecessors to `block`.
-    pub(crate) fn pred_iter(&self, basic_block_id: BasicBlockId) -> PredIter {
+    /// Get an iterator over the CFG predecessors to `basic_block_id`.
+    pub(crate) fn pred_iter(
+        &self,
+        basic_block_id: BasicBlockId,
+    ) -> impl ExactSizeIterator<Item = BasicBlockId> + '_ {
         self.data
             .get(&basic_block_id)
             .expect("ICE: Attempted to iterate predecessors of block not found within cfg.")
             .predecessors
             .iter()
+            .copied()
     }
 
-    /// Get an iterator over the CFG successors to `block`.
-    pub(crate) fn succ_iter(&self, basic_block_id: BasicBlockId) -> SuccIter {
+    /// Get an iterator over the CFG successors to `basic_block_id`.
+    pub(crate) fn succ_iter(
+        &self,
+        basic_block_id: BasicBlockId,
+    ) -> impl ExactSizeIterator<Item = BasicBlockId> + '_ {
         self.data
             .get(&basic_block_id)
             .expect("ICE: Attempted to iterate successors of block not found within cfg.")
             .successors
             .iter()
+            .copied()
     }
 }
 
-/// An iterator over block predecessors. The iterator type is `BasicBlockId`.
-pub(crate) type PredIter<'a> = hash_set::Iter<'a, BasicBlockId>;
+#[cfg(test)]
+mod tests {
+    use crate::ssa_refactor::ir::{instruction::TerminatorInstruction, types::Type};
 
-/// An iterator over block successors. The iterator type is `BasicBlockId`.
-pub(crate) type SuccIter<'a> = hash_set::Iter<'a, BasicBlockId>;
+    use super::{super::function::Function, ControlFlowGraph};
+
+    #[test]
+    fn empty() {
+        let mut func = Function::new();
+        let block_id = func.entry_block();
+        func.dfg[block_id].set_terminator(TerminatorInstruction::Return { return_values: vec![] });
+
+        ControlFlowGraph::with_function(&func);
+    }
+
+    #[test]
+    fn jumps() {
+        let mut func = Function::new();
+        let block0_id = func.entry_block();
+        let cond = func.dfg.add_block_parameter(block0_id, Type::unsigned(1));
+        let block1_id = func.dfg.new_block();
+        let block2_id = func.dfg.new_block();
+
+        func.dfg[block0_id].set_terminator(TerminatorInstruction::JmpIf {
```

Comment:
A comment here showing the structure of the ssa you're building could be helpful:
```
// fn func {
// block0(cond: u1):
//     jmpif cond(), then: block2, else: block1
// block1():
//     jmpif cond(), then: block1, else: block2
// block2():
//     return
// }
```